File: src/main/java/com/intellij/tapestry/core/util/chain/ChainBase.java
package com.intellij.tapestry.core.util.chain;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Base {@link Chain} implementation: runs its commands in order and stops at the
 * first one that returns {@code true}.
 */
public class ChainBase implements Chain {

    private final List<Command> commands = new ArrayList<>();

    public ChainBase() {
    }

    public ChainBase(Command[] commands) {
        this.commands.addAll(Arrays.asList(commands));
    }

    @Override
    public void addCommand(Command command) {
        commands.add(command);
    }

    @Override
    public boolean execute(Context context) throws Exception {
        for (Command command : commands) {
            if (command.execute(context)) {
                return true;
            }
        }
        return false;
    }
}

File: src/main/kotlin/com/github/rar91279/plugin/tapestry/core/util/chain/ChainBase.kt
package com.github.rar91279.plugin.tapestry.core.util.chain

/**
 * Base [Chain] implementation: runs its commands in order and stops at the
 * first one that returns `true`.
 */
open class ChainBase() : Chain {

    private val commands = mutableListOf<Command>()

    constructor(commands: Array<Command>) : this() {
        this.commands.addAll(commands)
    }

    override fun addCommand(command: Command) {
        commands.add(command)
    }

    @Throws(Exception::class)
    override fun execute(context: Context): Boolean = commands.any { it.execute(context) }
}
